### DISCUSSION TITLE
Confusion regarding the licensing of vscode-builtin-cpp and vscode-cpptools

### DISCUSSION THREAD

**Discussion by yunuscukran** [eclipse-theia/theia#8564] [General] [answered]
Hi folks,

I wanted to raise a potential problem related to [vscode-builtin-cpp]( https://open-vsx.org/extension/vscode/cpp), which I believe is the preferred way to add C/C++ functionality to Theia. While I was trying to figure out why PlatformIO's VSCode extension [can't](https://github.com/open-vsx/publish-extensions/issues/86) be put on the OpenVSX registry, I stumbled open this recent [discussion](https://github.com/microsoft/vscode-cpptools/issues/5784) regarding the licensing of VSCode's C/C++ extension.

This part of the conversation seemed worrying:
> The intent of the license(s) is that our binaries are only to be used by the Visual Studio family of products, while the TypeScript code is free to copy/use (MIT). **Which means that no, you may not use the binaries or the VSIX in Eclipse Theia or any other IDE.**

The [license](https://github.com/microsoft/vscode-cpptools/blob/master/Extension/LICENSE.txt) they mention is quite strict:
> MICROSOFT C/C++ EXTENSION FOR VISUAL STUDIO CODE
> 
> 1. INSTALLATION AND USE RIGHTS. You may only use the C/C++ Extension
>    for Visual Studio Code with Visual Studio Code, Visual Studio or
>    Xamarin Studio software to help you develop and test your
>    applications.

I wanted to confirm that the C++ functionality, which I get from [here](https://open-vsx.org/extension/vscode/cpp), does not use these mentioned binaries, and is safe regarding this issue. 

Thanks

**Comment by vince-fugnitto** [comment, marked as answer]
@yunuscukran thank you for the discussion, the two extensions you referenced are different (one being `vscode-builtin-cpp` and the other being `vscode-cpptools`). You are right that `vscode-cpptools` is quite restrictive, it's one of the reasons we do not use it ourselves, nor is it present in our example apps under [theia-apps](https://github.com/theia-ide/theia-apps). Instead, we use [`vscode-clangd`](https://github.com/clangd/vscode-clangd) extension maintained by the LLVM community.

The [`vscode-builtin-cpp`](https://github.com/microsoft/vscode/tree/master/extensions/cpp) like other builtins are part of vscode repository (safe to use) and it only provides texmate grammars and snippets for the C/C++ language (these extensions are published by the community under the [theia-ide/vscode-builtin-extensions](https://github.com/theia-ide/vscode-builtin-extensions) repository.

**Reply by yunuscukran** [reply]
It is mentioned that the code itself for `vscode-cpptools` is open source. Would it be possible to build that from the source and use it with Theia? We have another extension that depends on it...

**Reply by vince-fugnitto** [reply]
> It is mentioned that the code itself for `vscode-cpptools` is open source. Would it be possible to build that from the source and use it with Theia? We have another extension that depends on it...

@yunuscukran the best place to look would be the `vscode-cpptools` repository, since we are not lawyers and cannot confirm if the you can legally use the software for your own application (based on the description of their license it does not look possible outside of family of Microsoft products). You may want reference past discussions on their repository or clarify with the original authors. For instance, the following issue https://github.com/microsoft/vscode-cpptools/issues/2274 states that while the source code is open-source, the distributed binaries are not and are quite restrictive.

- https://github.com/microsoft/vscode-cpptools/issues/5784
- https://github.com/microsoft/vscode-cpptools/issues/2274

We do not use the `vscode-cpptools` in the repo or any example application due to the license declared [here](https://marketplace.visualstudio.com/items/ms-vscode.cpptools/license).

> INSTALLATION AND USE RIGHTS. a) General. You may install and use any number of copies of the software only with Microsoft Visual Studio, Visual Studio for Mac, Visual Studio Code, Azure DevOps, Team Foundation Server, and successor Microsoft products and services (collectively, the “Visual Studio Products and Services”) to develop and test your applications.

**Reply by yunuscukran** [reply]
Thank you for the detailed response.

When I tried to explore the `vscode-clang `option instead, I got the same compatibility issue as https://github.com/eclipse-theia/theia/discussions/8601, (with Theia 1.5.0). I couldn't determine from that conversation whether Theia 1.5 currently supports this extension or not,

**Reply by vince-fugnitto** [reply]
> Thank you for the detailed response.
> 
> When I tried to explore the `vscode-clang `option instead, I got the same compatibility issue as #8601, (with Theia 1.5.0). I couldn't determine from that conversation whether Theia 1.5 currently supports this extension or not,

@yunuscukran you can either downgrade the extension to a compatible version (which is what the [`theia-cpp-docker`](https://github.com/theia-ide/theia-apps/blob/b86d86ba43f4c7af87d8a5f4a3c5ea63ef881731/theia-cpp-docker/latest.package.json#L121)) uses, or update the default vscode api version which is a simple check based on the extension's `engines.vscode` version.
